cmd/clusterlint/main.go

```diff
@@ -133,7 +135,17 @@ func listChecks(c *cli.Context) error {
 
 // runChecks runs all the checks based on the flags passed.
 func runChecks(c *cli.Context) error {
-	client, err := kube.NewClient(kube.WithConfigFile(c.GlobalString("kubeconfig")), kube.WithKubeContext(c.GlobalString("context")), kube.WithTimeout(c.GlobalDuration("timeout")))
+	var kubeconfigFilePaths []string
+
+	if kubeconfig := c.GlobalString("kubeconfig"); kubeconfig != "" {
+		kubeconfigFilePaths = []string{kubeconfig}
+	} else if value := os.Getenv("KUBECONFIG"); value != "" {
+		kubeconfigFilePaths = strings.Split(value, delimiter)
+	} else {
+		kubeconfigFilePaths = []string{filepath.Join(os.Getenv("HOME"), ".kube", "config")}
+	}
```

Comment:
I believe this logic is basically covered by [`clientcmd.NewDefaultClientConfigLoadingRules()`](https://github.com/digitalocean/clusterlint/blob/f0cfce9e4bb59b7d48cbd647c20d22a100b88606/vendor/k8s.io/client-go/tools/clientcmd/loader.go#L137) including proper merging and more -- for instance, I think `$HOME` does not apply under Windows. The `clientcmd` package handles that appropriately.

I think the [Velero code base uses a configuration](https://github.com/heptio/velero/blob/b8f3a008cb6aa15df417e4d0e0c5620ef6c13132/pkg/client/client.go#L33-L37) that matches what we want too.

(Note that I haven't tested this, maybe you can check that.)

Reply:
@timoreimann, @adamwg  - earlier, we used to have what the Velero codebase has. We changed that to set `Precendence` instead of `ExplicitPath` to support the fact that multiple config file paths can be provided in the KUBECONFIG env var. 

With respect to the default value, the [documentation](https://godoc.org/k8s.io/client-go/tools/clientcmd#BuildConfigFromFlags) mentions that the default config will be used in case the kubeconfig file path is not specified. This is in accordance with the config loading rules for kubectl (see kubectl config -h). I removed the above line and tried doing -

```
if len(defOpts.paths) != 0 {
		loadingRules := &clientcmd.ClientConfigLoadingRules{Precedence: defOpts.paths}
		configOverrides := &clientcmd.ConfigOverrides{}
		if defOpts.kubeContext != "" {
			configOverrides.CurrentContext = defOpts.kubeContext
		}
		config, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides).ClientConfig()
	} else {
                 // for default paths
		// config, err = clientcmd.BuildConfigFromFlags("", "")
                //config, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{}, &clientcmd.ConfigOverrides{}).ClientConfig()
	}
```
The error I get is `invalid configuration: no configuration has been provided`.  

I think we should retain the code as is right now. At least the default can be honoured for Linux and Mac users. Windows users can specify the kubeconfig path or env var explicitly? WDYT?

Reply:
>earlier, we used to have what the Velero codebase has. We changed that to set Precendence instead of ExplicitPath to support the fact that multiple config file paths can be provided in the KUBECONFIG env var.

Multiple config files specified in the KUBECONFIG file seem to also be parsed when using `clientcmd.NewDefaultClientConfigLoadingRules()` (precisely [here](clientcmd.NewDefaultClientConfigLoadingRules())), and they then become part of the precedence chain. I think the idea of `ExplicitPath` is to indicate that the user wants a particular kubeconfig for sure -- that's why the client builder returns an error if that file is missing (even though other files from the KUBECONFIG environment variable might still work).
So it sounds like what we want is already provided. What am I missing?